crates/circuit/src/operations.rs

```diff
@@ -2103,6 +2118,24 @@ impl Operation for PyInstruction {
     fn control_flow(&self) -> bool {
         self.control_flow
     }
+    fn blocks(&self) -> Vec<CircuitData> {
+        if !self.control_flow {
+            return vec![];
+        }
+        Python::with_gil(|py| -> Vec<CircuitData> {
+            let raw_blocks = self.instruction.getattr(py, "blocks").unwrap();
+            let blocks: &Bound<PyTuple> = raw_blocks.downcast_bound::<PyTuple>(py).unwrap();
+            blocks
+                .iter()
+                .map(|b| {
+                    b.getattr(intern!(py, "_data"))
+                        .unwrap()
+                        .extract::<CircuitData>()
+                        .unwrap()
+                })
+                .collect()
+        })
```

Comment:
```suggestion
        Python::with_gil(|py| -> PyResult<Vec<CircuitData>> {
            let raw_blocks = self.instruction.getattr(py, "blocks")?;
            let blocks: &Bound<PyTuple> = raw_blocks.downcast_bound::<PyTuple>(py)?;
            blocks
                .iter()
                .map(|b| -> PyResult<_> {
                    b.getattr(intern!(py, "_data"))?.extract::<CircuitData>()
                })
                .collect::<PyResult<_>>()
        })
        .unwrap_or_default()
```
You might find this option more attractive. We let the iterator and the `Python::with_gil()` do their own result handling and we only unwrap at the end and use the `Default` option were it to ever fail. This option is more stable as it has zero chance of ever panicking, if it ever were to fail an empty list will be returned.

Reply:
This is actually very similar to what I had originally in 622509eb3e48f6a0c6bfa8f13cf7686b94c842c5 🙂 (the difference being that I was returning an `Option` instead of an empty `Vec` if something went wrong).

In the current version, I actually purposefully made this code less flexible. The reason being that if one of these expectations is not true (e.g. if `self.instruction` does not have a `blocks` attribute), then it means that there is a _**bug**_ in Qiskit. In that case, returning an empty `Vec` would only mask the underlying issue and prevent us from finding out about it / fixing it.

Reply:
Fwiw I'm fairly sure I've done the same thing as Kevin elsewhere and taken the `control_flow` attribute to mean that it's an instance of `ControlFlowOp` and therefore `blocks` must exist (which matches the Python expectations).

Reply:
That's good then, can you add a comment explaining this? I think it's a good idea to mention that all of this unwrapping is being done with the purpose of catching bad behavior.

Reply:
Good idea. Done in 2a0951fd5201f4890bdb2213cb23404c77eaee17.